### PR TITLE
feat: add retry directive

### PR DESCRIPTION
**Pull request by matt2e**
adds retry directive for fsm transition verbs. This PR does not contain logic to execute retries.
- go expects format `//ftl:retry [count] minbackoff [maxbackoff]`
- schema has format `+retry [count] minbackoff [maxbackoff]`
- backoffs support `s`, `m`, `h`, `d` units like `90s` or `1m30s`
- validation makes sure retry directive is only added to verbs that are part of a FSM within the same module
- min/max backoffs are `1s`/`1d` (very debatable...)

### CODE REVIEW

File: backend/schema/metadataretry.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"strings"
+	"time"
+
+	"github.com/alecthomas/types/optional"
+	"google.golang.org/protobuf/proto"
+
+	schemapb "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/schema"
+)
+
+const (
```
Comment (matt2e) [inline backend/schema/metadataretry.go:16]: String constants are just to make error printing have the same format we expect users to use (ie `1d` not go's standard `24h0m0s`

File: backend/schema/validate_test.go [hunk 2]
```diff
@@ -291,7 +293,74 @@ func TestValidate(t *testing.T) {
...
+						transition one.A to one.E
+						transition one.A to one.F
+						transition one.A to one.G
+						transition one.A to one.H
+						transition one.A to one.I
+						transition one.A to one.J
+					}
+				}
+				`,
+			errs: []string{
```
Comment (matt2e) [inline backend/schema/validate_test.go:353]: Please read these to see if they are clear

File: backend/schema/metadataretry.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		components = append(components, m.MaxBackoff)
+	}
+	return strings.Join(components, " ")
+}
+
+func (m *MetadataRetry) ToProto() proto.Message {
+	count := int64(*m.Count)
+	return &schemapb.MetadataRetry{
+		Pos:        posToProto(m.Pos),
+		Count:      &count,
```
Comment (alecthomas) [inline backend/schema/metadataretry.go:52]: You should be able to use `proto.Int64(int64(m.Count))` here.

File: backend/schema/metadataretry.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+func parseRetryDuration(str string) (time.Duration, error) {
+	// regex is more lenient than what is valid to allow for better error messages.
+	re := regexp.MustCompile(`(\d+)([a-zA-Z]+)`)
+
+	var duration time.Duration
+	previousUnitDuration := time.Duration(0)
+	for len(str) > 0 {
+		matches := re.FindStringSubmatchIndex(str)
+		if matches == nil {
+			return 0, fmt.Errorf("unable to parse retry backoff %q. Expected duration in format like '1m' or '30s'", str)
```
Comment (alecthomas) [inline backend/schema/metadataretry.go:89]: In general try not to make Go errors include sentences, as when they're chained it makes them read quite awkwardly. Instead prefer to use semi-colons, hyphens, or commas: `unable to parse retry backoff %q - expected duration in format like ...`

File: backend/schema/metadataretry.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		case "d":
+			unitDuration = time.Hour * 24
+		case "h":
+			unitDuration = time.Hour
+		case "m":
+			unitDuration = time.Minute
+		case "s":
+			unitDuration = time.Second
+		default:
+			return 0, fmt.Errorf("retry has unknown unit %q. Use 'd', 'h', 'm' or 's'", unitStr)
```
Comment (alecthomas) [inline backend/schema/metadataretry.go:108]: Same here as above.

File: backend/schema/metadataretry.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		}
+		if previousUnitDuration != 0 && previousUnitDuration <= unitDuration {
+			return 0, fmt.Errorf("retry has unit %q out of order. Units need to be ordered from largest to smallest", unitStr)
```
Comment (alecthomas) [inline backend/schema/metadataretry.go:111]: Same here as above.

File: backend/schema/metadataretry.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	MaxBackoffLimitStr = "1d"
+	MaxBackoffLimit    = 24 * time.Hour
+)
+
+type MetadataRetry struct {
+	Pos Position `parser:"" protobuf:"1,optional"`
+
+	Count      *int   `parser:"'+' 'retry' (@Number Whitespace)?" protobuf:"2,optional"`
+	MinBackoff string `parser:"@(Number Ident)?" protobuf:"3"`
+	MaxBackoff string `parser:"@(Number Ident)?" protobuf:"4"`
```
Comment (matt2e) [inline backend/schema/metadataretry.go:28]: FYI, the reasoning behind `@(Number Ident)?` is that I could not find a sane way to parse:
- multiple number + unit pairs while ensuring no whitespace inbetween
- `1m30s` is tokenized as `1` (number) + `m30s` (ident)
Instead we are just getting these in as strings and then parsing them ourselves using regex. This helps us customize the error messages for more cases.

File: backend/schema/validate.go [hunk 8]
```diff
@@ -521,6 +522,52 @@ func validateVerbMetadata(scopes Scopes, module *Module, n *Verb) (merr []error)
...
+				merr = append(merr, errorf(md, "verb %s: retry count must be atleast 1", n.Name))
+			}
+
+			// Validate parsing of durations
+			minDuration, err := md.MinBackoffDuration()
+			if err != nil {
+				merr = append(merr, errorf(md, "verb %s: %v", n.Name, err))
+				return
+			}
+			if minDuration == 0 {
```
Comment (alecthomas) [inline backend/schema/validate.go:559]: Should this check for `<time.Second` / `MinBackoffDuration`?

File: backend/schema/validate.go [hunk 9]
```diff
@@ -521,6 +522,52 @@ func validateVerbMetadata(scopes Scopes, module *Module, n *Verb) (merr []error)
...
+				merr = append(merr, errorf(md, "verb %s: retry duration must be atleast 1s", n.Name))
+				return
+			}
+			maxDuration, err := md.MaxBackoffDuration()
+			if err != nil {
+				merr = append(merr, errorf(md, "verb %s: %v", n.Name, err))
+				return
+			}
+			if maxDuration, ok := maxDuration.Get(); ok && maxDuration < minDuration {
```
Comment (alecthomas) [inline backend/schema/validate.go:568]: Should there be a check against `MaxBackoffDuration` (ie. 24h)?

File: backend/schema/validate_test.go [hunk 10]
```diff
@@ -291,7 +294,78 @@ func TestValidate(t *testing.T) {
...
+				`12:7-7: verb E: retry must have a minimum backoff`,
+				`14:7-7: verb F: retry has unit "m" out of order. Units need to be ordered from largest to smallest`,
+				`17:7-7: verb can not have multiple instances of retry`,
+				`19:7-7: verb H: retry has unknown unit "mins". Use 'd', 'h', 'm' or 's'`,
+				`21:7-7: verb I: max backoff duration (1s) needs to be atleast as long as initial backoff (1m)`,
+				`23:7-7: verb J: retry backoff can not be larger than 1d`,
+				`25:7-7: verb K: retry count must be atleast 1`,
+				`4:7-7: verb A: retry has unit "m" out of order. Units need to be ordered from largest to smallest`,
+				`6:7-7: verb B: retry has unit "d" out of order. Units need to be ordered from largest to smallest`,
+				`8:7-7: verb C: retry must have a minimum backoff of 1s`,
```
Comment (alecthomas) [inline backend/schema/validate_test.go:368]: Nice

File: go-runtime/compile/parser.go [hunk 11]
```diff
@@ -130,12 +131,35 @@ func (d *directiveCronJob) String() string {
 	return fmt.Sprintf("cron %s", d.Cron)
 }
 
+type directiveRetry struct {
+	Pos schema.Position
+
+	Count      *int   `parser:"'retry' (@Number Whitespace)?"`
+	MinBackoff string `parser:"@(Number Ident)"`
```
Comment (alecthomas) [inline go-runtime/compile/parser.go:138]: I think you'll need something like `@(Number (?! Whitespace) Ident)` here to catch cases like `1 m`

**Review by alecthomas — APPROVED** [review]
Nice nice. Will we add FSM-wide retries (ie. a retry annotation on the FSM itself) in a future PR?

File: backend/schema/metadataretry.go (hunk 3 above)
Comment (matt2e) [reply backend/schema/metadataretry.go:52]: Oh, looks like i made a more fundamental mistake... not handling `m.Count == 0`
Updated to
```go
var count *int64
if m.Count != nil {
  count = proto.Int64(int64(*m.Count))
}
```

File: backend/schema/validate.go (hunk 8 above)
Comment (matt2e) [reply backend/schema/validate.go:559]: Ah, removed as it wasn't ever hit anymore. It's now centralized in MetadataRetry.go

File: backend/schema/validate.go (hunk 9 above)
Comment (matt2e) [reply backend/schema/validate.go:568]: Both MinBackoff and MaxBackoff are checked for min/max during `parseRetryDuration`

File: go-runtime/compile/parser.go (hunk 11 above)
Comment (matt2e) [reply go-runtime/compile/parser.go]: Done, and added a test for it 👍

**Pull request merged**